activerecord/lib/active_record/connection_adapters/abstract/query_log_tags.rb

```diff
@@ -0,0 +1,240 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/module/attribute_accessors_per_thread"
+
+module ActiveRecord
+  module ConnectionAdapters
+    module QueryLogTags
+      extend ActiveSupport::Concern
+      included do
+        mattr_accessor :prepend_comment, default: false
+      end
+
+      module ClassMethods
+        def prepend_execution_methods # :nodoc:
+          descendants.each do |klass|
+            # Prepend execution methods for edge descendants of AbstractAdapter
+            klass.prepend(ExecutionMethods) if klass.descendants.empty?
+          end
+        end
+
+        def add_query_log_tags_to_sql(sql) # :nodoc:
+          return sql unless QueryLogTagsContext.tags_available?
+          comments = [QueryLogTagsContext.comment, QueryLogTagsContext.inline_comment].compact
+          comments.each do |comment|
+            if comment.present? && !sql.include?(comment)
+              sql = if prepend_comment
+                "#{comment} #{sql}"
+              else
+                "#{sql} #{comment}"
+              end
+            end
+          end
```

Comment:
This spills too much of the Context class into here. Let's extract a `comments` helper we can use here:

```suggestion
          def add_query_log_tags_to_sql(sql) # :nodoc:
            QueryLogTagsContext.comments.each do |comment|
              unless sql.include?(comment)
                sql = prepend_comment ? "#{comment} #{sql}" : "#{sql} #{comment}"
              end
            end
```

Then in the context we'd inline `tags_available?` and do something like:

```ruby
def comments
  [ comment if components.any?, inline_comment if inline_annotations.any? ].reject(&:blank?)
end
```